experimental/gradually-typed/src/Torch/GraduallyTyped/Examples/TwoLayerNetwork.hs

```diff
@@ -0,0 +1,407 @@
+  NamedModel
+    ( TwoLayerNetwork
+        (FstLayerF gradient device dataType inputDim hiddenDim)
+        (SndLayerF gradient device dataType outputDim hiddenDim)
+    )
+
+-- | Specifies the first layer of the neural network.
+type FstLayerF gradient device dataType inputDim hiddenDim =
+  NamedModel
+    ( GLinear
+        (NamedModel (LinearWeightF gradient device dataType inputDim hiddenDim))
+        (NamedModel (LinearBiasF 'WithBias gradient device dataType hiddenDim))
+    )
+
+-- | Specifies the second layer of the neural network.
+type SndLayerF gradient device dataType outputDim hiddenDim =
+  NamedModel
+    ( GLinear
+        (NamedModel (LinearWeightF gradient device dataType hiddenDim outputDim))
+        (NamedModel (LinearBiasF 'WithBias gradient device dataType outputDim))
+    )
+
+-- | Specify the parameters of the two-layer neural network
+twoLayerNetworkSpec ::
+  forall gradient device dataType inputDim outputDim hiddenDim.
+  -- | whether or not to compute gradients for the parametrs
+  SGradient gradient ->
+  -- | which device to use
+  SDevice device ->
+  -- | which data type to use
+  SDataType dataType ->
+  -- | input dimension
+  SDim inputDim ->
+  -- | output dimension
+  SDim outputDim ->
+  -- | hidden dimension
+  SDim hiddenDim ->
+  -- | specification for the network
+  ModelSpec (TwoLayerNetworkF gradient device dataType inputDim outputDim hiddenDim)
+twoLayerNetworkSpec gradient device dataType inputDim outputDim hiddenDim =
+  NamedModel "twoLayerNetwork" $
+    TwoLayerNetwork
+      (NamedModel "fstLayer" $ linearSpec SWithBias gradient device dataType inputDim hiddenDim)
+      (NamedModel "sndLayer" $ linearSpec SWithBias gradient device dataType hiddenDim outputDim)
+
+-- | 'HasForward' instance used to define the forward pass of the model
+instance
+  ( HasForward
+      fstLayer
+      (Tensor gradient layout dataType device shape)
+      generatorDevice
+      (Tensor gradient0 layout0 dataType0 device0 shape0)
+      generatorDevice0,
+    HasForward
+      sndLayer
+      (Tensor gradient0 layout0 dataType0 device0 shape0)
+      generatorDevice0
+      output
+      generatorOutputDevice
+  ) =>
+  HasForward
+    (TwoLayerNetwork fstLayer sndLayer)
+    (Tensor gradient layout dataType device shape)
+    generatorDevice
+    output
+    generatorOutputDevice
+  where
+  forward TwoLayerNetwork {..} input =
+    runIxStateT $
+      ireturn input
+        >>>= IxStateT . forward fstLayer
+        >>>= ireturn . Torch.GraduallyTyped.tanh
+        >>>= IxStateT . forward sndLayer
+
+-- | Options for the Adam optimizer.
+data AdamOptions = AdamOptions
+  { -- | learning rate
+    learningRate :: Double,
+    -- | beta1
+    beta1 :: Double,
+    -- | beta2
+    beta2 :: Double,
+    -- | epsilon
+    epsilon :: Double,
+    -- | weight decay
+    weightDecay :: Double,
+    -- | use amsgrad
+    amsgrad :: Bool
+  }
```

Comment:
Would you like to create Torch.GraduallyTyped.Optim later and include it?

Reply:
Yes, at some point